lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java

```diff
@@ -337,9 +354,79 @@ public MavenPluginManager getMavenPluginManager() {
 	}
 
 	public void didChangeWorkspaceFolders(URI[] added, URI[] removed) {
+		initialize();
 		WorkspaceReader workspaceReader = mavenRequest.getWorkspaceReader();
 		if (workspaceReader instanceof MavenLemminxWorkspaceReader) {
-			// TODO
+			Collection<DOMDocument> projectsToAdd = computeAddedWorkspaceProjects(added != null? added : new URI[0]);
+			Collection<URI> projectsToRemove = computeRemovedWorkspaceProjects(removed != null ? removed : new URI[0]);
+
+			if (projectsToAdd != null) {
+				projectsToAdd.stream().forEach(d -> getProjectCache().addDocument(URI.create(d.getDocumentURI()), d));
+			}
+
+			if (projectsToRemove != null) {
+				projectsToRemove.stream().forEach(u -> getProjectCache().removeDocument(u));
+			}
 		}
 	}
+	
+	private Collection<URI> computeRemovedWorkspaceProjects(URI[] removed) {
+		Collection<MavenProject> cachedProjects = getProjectCache().getProjects();
+		return cachedProjects.stream().filter(p -> {
+			return Arrays.asList(removed).stream().filter(uri -> {
+				Path removedPath = new File(uri).toPath();
+				return p.getFile().toPath().startsWith(removedPath);
+			}).findAny().isPresent();
+		}).map(p -> p.getFile().toURI()).collect(Collectors.toUnmodifiableList());
+	}
+	
+	private List<DOMDocument> computeAddedWorkspaceProjects(URI[] added) {
+		List<DOMDocument> projectsToAdd = new ArrayList<DOMDocument>();
+		Arrays.asList(added).stream().forEach(uri -> {
+			Path addedPath = new File(uri).toPath();
+			try {
+				Files.walkFileTree(addedPath, Collections.emptySet(), 10, new SimpleFileVisitor<Path>() {
+					@Override
+					public FileVisitResult preVisitDirectory(Path file, BasicFileAttributes attrs) throws IOException {
+						if (file.getFileName().toString().charAt(0) == '.') {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+						return FileVisitResult.CONTINUE;
+					}
+
+					@Override
+					public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+						if (file.getFileName().toString().toLowerCase().endsWith(".xml")) {
+							DOMDocument document;
+							try {
+								document = createDOMDocument(file.toUri());
+								if (document != null) {
+									DOMElement documentElement = document.getDocumentElement();
+									if (documentElement != null && PROJECT_ELT.equals(documentElement.getLocalName())) {
+										projectsToAdd.add(document);
+									}
+								}
+							} catch (URISyntaxException e) {
+								LOGGER.log(Level.SEVERE, e.getMessage(), e);
+							}
+						}
+						return FileVisitResult.CONTINUE;
+					}
+				});
+			} catch (IOException e) {
+				LOGGER.log(Level.SEVERE, e.getMessage(), e);
+			}
+		});
+		
+		return projectsToAdd;
+	}
+	
+	private DOMDocument createDOMDocument(URI uri) throws IOException, URISyntaxException {

```

Comment:
Couldn't this be part of ProjectCache instead?

Reply:
`visitFile()` uses this method for filtering purposes (in order to collect only xml files with `<project/>` root element. Also, once a document is already created here it  is passed to `MavenProjectCache.addDocument(URI, DOMDocument)` method.

Reply:
I think we should just rely on filename at this stage and not look at document content here.

Reply:
Yes, but I have moved that root element check to addDocument(), so if a document doesn't have `<project />` element then it's just not added to the cache. This way it works much faster.